rules/aip0136/response_message_name.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aip0136
+
+import (
+	"fmt"
+
+	"github.com/googleapis/api-linter/lint"
+	"github.com/googleapis/api-linter/locations"
+	"github.com/googleapis/api-linter/rules/internal/utils"
+	"github.com/jhump/protoreflect/desc"
+)
+
+// Custom methods should return a response message matching the RPC name,
+// with a Response suffix, or the resource being operated on.
+var responseMessageName = &lint.MethodRule{
+	Name:   lint.NewRuleName(136, "response-message-name"),
+	OnlyIf: isCustomMethod,
+	LintMethod: func(m *desc.MethodDescriptor) []lint.Problem {
+		// A response is considered valid if
+		// - The response name matches the RPC name with a `Response` suffix
+		// - The response is the resource being operated on
+		// To identify the resource being operated on, we inspect the Resource
+		// Reference of the input type's `name` field. This guidance is documented
+		// in https://google.aip.dev/136#resource-based-custom-methods
+
+		response := utils.GetResponseType(m)
+		requestType := utils.GetResourceReference(m.GetInputType().FindFieldByName("name")).GetType()
+		resourceOperatedOn := utils.FindResourceMessage(requestType, m.GetFile())
+
+		// Output type has `Response` suffix
+		if len(utils.LintMethodHasMatchingResponseName(m)) == 0 {
+			return nil
+		}
```

Comment:
I'd move this above the resource-related stuff to short circuit the rule. Note the saving of the suffix-based findings for later...

```suggestion
    // Output type has `Response` suffix
    suffixFindings := utils.LintMethodHasMatchingResponseName(m)
		if len(suffixFindings) == 0 {
			return nil
		}

		response := utils.GetResponseType(m)
		requestType := utils.GetResourceReference(m.GetInputType().FindFieldByName("name")).GetType()
		resourceOperatedOn := utils.FindResourceMessage(requestType, m.GetFile())
```